src/develop/lightroom.c

```diff
@@ -469,560 +1009,109 @@ void dt_lightroom_import(int imgid, dt_develop_t *dev, gboolean iauto)
-        }
-        tagNode = tagNode->next;
-      }
-    }
-    else if(dev != NULL && !xmlStrcmp(entryNode->name, (const xmlChar *)"RetouchInfo"))
-    {
-      xmlNodePtr riNode = entryNode;
+  // record the name-spaces needed for the parsing
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "crs",
+     BAD_CAST "http://ns.adobe.com/camera-raw-settings/1.0/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "dc",
+     BAD_CAST "http://purl.org/dc/elements/1.1/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "tiff",
+     BAD_CAST "http://ns.adobe.com/tiff/1.0/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "xmp",
+     BAD_CAST "http://ns.adobe.com/xap/1.0/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "exif",
+     BAD_CAST "http://ns.adobe.com/exif/1.0/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "lr",
+     BAD_CAST "http://ns.adobe.com/lightroom/1.0/");
+  xmlXPathRegisterNs
+    (xpathCtx,
+     BAD_CAST "rdf",
+     BAD_CAST "http://www.w3.org/1999/02/22-rdf-syntax-ns#");
 
-      riNode = riNode->xmlChildrenNode;
-      riNode = riNode->next;
-      riNode = riNode->xmlChildrenNode;
-      riNode = riNode->next;
+  // All prefixes to parse from the XMP document
+  static char *names[] = { "crs", "dc", "tiff", "xmp", "exif", "lr", NULL };
 
-      while(riNode)
-      {
-        if(!xmlStrcmp(riNode->name, (const xmlChar *)"li"))
-        {
-          xmlChar *value = xmlNodeListGetString(doc, riNode->xmlChildrenNode, 1);
-          spot_t *p = &ps.spot[ps.num_spots];
-          if(sscanf((const char *)value, "centerX = %f, centerY = %f, radius = %f, sourceState = %*[a-zA-Z], "
-                                         "sourceX = %f, sourceY = %f",
-                    &(p->x), &(p->y), &(p->radius), &(p->xc), &(p->yc)))
-          {
-            ps.num_spots++;
-            has_spots = TRUE;
-          }
-          xmlFree(value);
-        }
-        if(ps.num_spots == MAX_SPOTS) break;
-        riNode = riNode->next;
-      }
-    }
-    else if(dev != NULL && !xmlStrcmp(entryNode->name, (const xmlChar *)"ToneCurvePV2012"))
+  for (int i=0; names[i]!=NULL; i++)
     {
-      xmlNodePtr tcNode = entryNode;
+      char expr[50];
 
-      tcNode = tcNode->xmlChildrenNode;
-      tcNode = tcNode->next;
-      tcNode = tcNode->xmlChildrenNode;
-      tcNode = tcNode->next;
-
-      while(tcNode)
-      {
-        if(!xmlStrcmp(tcNode->name, (const xmlChar *)"li"))
-        {
-          xmlChar *value = xmlNodeListGetString(doc, tcNode->xmlChildrenNode, 1);
+      /* Lr 7.0 CC (nodes) */
+      snprintf(expr, sizeof(expr), "//%s:*", names[i]);
+      _handle_xpath(dev, doc, imgid, xpathCtx, (const xmlChar *)expr, &data);
 
-          if(sscanf((const char *)value, "%d, %d", &(curve_pts[n_pts][0]), &(curve_pts[n_pts][1]))) n_pts++;
-          xmlFree(value);
-        }
-        if(n_pts == MAX_PTS) break;
-        tcNode = tcNode->next;
-      }
+      /* Lr up to 6.0 (attributes) */
```

Comment:
Can a Lr Xmp contain both 6.0 attributes and newer ones? If yes, it might make sense to check the old ones first and let the new ones overwrite them afterwards.
If they are mutually exclusive then ignore this comment.

Reply:
No, this won't happen. The new XMP I've seen are fully converted to new format.